internal/docstore/query.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package docstore
+
+import (
+	"context"
+	"io"
+
+	"gocloud.dev/internal/docstore/driver"
+)
+
+// Query represents a query over a collection.
+type Query struct {
+	coll *Collection
+	dq   *driver.Query
+}
+
+// Query creates a new Query over the collection.
+func (c *Collection) Query() *Query {
+	return &Query{coll: c, dq: &driver.Query{}}
+}
+
+// Where expresses a condition on the query.
+// Valid ops are: "=", ">", "<", ">=", "<=".
+func (q *Query) Where(field, op string, value interface{}) *Query {
+	q.dq.Filters = append(q.dq.Filters, driver.Filter{
+		Field: field,
+		Op:    op,
+		Value: value,
+	})
+	return q
+}
+
+// Limit will limit the results to at most n documents.
+func (q *Query) Limit(n int) *Query {
+	q.dq.Limit = n
+	return q
+}
+
+// Get returns an iterator for retrieving the documents specified by the query. If
+// field paths are provided, only those paths are set in the resulting documents.
+//
+// Call Stop on the iterator when finished.
+func (q *Query) Get(ctx context.Context, fieldpaths ...string) *DocumentIterator {
+	_ = q.coll.driver.RunQuery(ctx, q.dq)
+	return &DocumentIterator{iter: q.dq.Iter}
```

Comment:
Instead of passing the iterator back through a field in the Query, could it be the first return value?

Reply:
Also, we haven't talked about how to handle delete and update queries in the driver. What about three separate methods?
```
RunGetQuery(ctx, query) (DocumentIterator, error)
RunUpdateQuery(ctx, query) error
RunDeleteQuery(ctx, query) error
```

Reply:
Sorry, that should be
```
RunUpdateQuery(ctx, query, mods) error
```

Reply:
I was originally plan to have the `RunQuery` to be similar to `RunActions`, where the `Action` is a field of `driver.Query`, which is also the reason I pass the iterator back through a field. I feel that seems more consistent and simpler for the driver interface.

Is there any other reason to have three separate methods here?

Reply:
No, it was just because of the different signatures.

I'm okay with using an enum but I'd still rather put the iterator in the return values. It's odd to pass something back in an argument. I know that for `Get` we populate the argument doc, but we need the arg to do type-driven decoding, and it's also forced by the fluent API. Neither of those things are at play here.

Reply:
Changed to `RunGetQuery`. I think returning an iterator looks good.